File: web/src/app/core/country/country-scope.service.ts
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Injectable, computed, inject, signal } from '@angular/core';
import { firstValueFrom } from 'rxjs';

import { AuthService } from '@/core/auth/auth.service';
import { ApiPaths } from '@/core/constants/api-paths';
import { CountryBrief, parseCountryBrief } from '@/models/organization';
import { environment } from '../../../environments/environment';

const STORAGE_KEY = 'vaxiil.country';
export const RESOLVED_COUNTRY_HEADER = 'X-Resolved-Country';

interface StoredCountry {
  id: string;
  isoCode2: string;
  name?: string;
  flag?: string | null;
}

@Injectable({ providedIn: 'root' })
export class CountryScopeService {
  private readonly http = inject(HttpClient);
  private readonly auth = inject(AuthService);

  private readonly countriesCache = signal<CountryBrief[]>([]);
  private initPromise: Promise<void> | null = null;

  readonly country = signal<CountryBrief | null>(this.readStored());
  readonly countryId = computed(() => this.country()?.id ?? '');
  readonly isoCode2 = computed(() => {
    const code = this.country()?.isoCode2?.trim() ?? '';
    return code ? code.toUpperCase() : '';
  });

  browserTimezone(): string {
    try {
      return Intl.DateTimeFormat().resolvedOptions().timeZone || '';
    } catch {
      return '';
    }
  }

  setCountry(country: CountryBrief): void {
    this.country.set(country);
    this.writeStored(country);
  }

  setCountryById(countryId: string, countries?: CountryBrief[]): void {
    const list = countries?.length ? countries : this.countriesCache();
    const match = list.find((c) => c.id === countryId);
    if (match) {
      this.setCountry(match);
    }
  }

  rememberCountries(countries: CountryBrief[]): void {
    this.countriesCache.set(countries);
  }

  /** Bootstrap once: localStorage → profile default → geo-country → first country. */
  ensureInitialized(countries: CountryBrief[] = []): Promise<void> {
    if (countries.length) {
      this.rememberCountries(countries);
    }
    if (!this.initPromise) {
      this.initPromise = this.bootstrap(countries);
    }
    return this.initPromise;
  }

  /** When storage empty, persist backend-resolved ISO2 if we can map it. */
  hydrateFromResolvedHeader(iso2: string | null | undefined): void {
    if (this.country() || !iso2) {
      return;
    }
    const code = iso2.trim().toUpperCase();
    if (code.length !== 2) {
      return;
    }
    const match = this.countriesCache().find(
      (c) => c.isoCode2.trim().toUpperCase() === code,
    );
    if (match) {
      this.setCountry(match);
    }
  }

  private async bootstrap(countries: CountryBrief[]): Promise<void> {
    const list = countries.length ? countries : this.countriesCache();
    const stored = this.country();
    if (stored) {
      if (list.length) {
        const match = list.find((c) => c.id === stored.id) ??
          list.find((c) => c.isoCode2.toUpperCase() === stored.isoCode2.toUpperCase());
        if (match) {
          this.setCountry(match);
          return;
        }
      }
      return;
    }

    const preferred = this.auth.currentUser()?.defaultCountryId;
    if (preferred && list.length) {
      const match = list.find((c) => c.id === preferred);
      if (match) {
        this.setCountry(match);
        return;
      }
    }

    const detected = await this.fetchGeoCountry();
    if (detected) {
      this.setCountry(detected);
      return;
    }

    if (list[0]) {
      this.setCountry(list[0]);
    }
  }

  private async fetchGeoCountry(): Promise<CountryBrief | null> {
    try {
      const response = await firstValueFrom(
        this.http.get<Record<string, unknown> | null>(
          `${environment.apiBaseUrl}${ApiPaths.organizationsGeoCountry}`,
          {
            observe: 'response',
            headers: {
              'X-Timezone': this.browserTimezone(),
            },
          },
        ),
      );
      if (response.status === 204 || response.body == null) {
        this.captureResolvedHeader(response);
        return null;
      }
      const country = parseCountryBrief(response.body);
      this.captureResolvedHeader(response);
      return country;
    } catch {
      return null;
    }
  }

  private captureResolvedHeader(response: HttpResponse<unknown>): void {
    const iso = response.headers.get(RESOLVED_COUNTRY_HEADER);
    if (iso && !this.country()) {
      this.hydrateFromResolvedHeader(iso);
    }
  }

  private readStored(): CountryBrief | null {
    try {
      const raw = localStorage.getItem(STORAGE_KEY);
      if (!raw) {
        return null;
      }
      const parsed = JSON.parse(raw) as StoredCountry;
      if (!parsed?.id || !parsed?.isoCode2) {
        return null;
      }
      return {
        id: String(parsed.id),
        isoCode2: String(parsed.isoCode2).toUpperCase(),
        name: parsed.name ?? '',
        flag: parsed.flag ?? null,
        phoneCode: null,
      };
    } catch {
      return null;
    }
  }

  private writeStored(country: CountryBrief): void {
    const payload: StoredCountry = {
      id: country.id,
      isoCode2: country.isoCode2.toUpperCase(),
      name: country.name,
      flag: country.flag,
    };
    localStorage.setItem(STORAGE_KEY, JSON.stringify(payload));
  }
}

File: mobile/lib/core/country/country_scope_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_timezone/flutter_timezone.dart';
import 'package:vaxiil_mobile/core/constants/app_constants.dart';
import 'package:vaxiil_mobile/core/network/dio_client.dart';
import 'package:vaxiil_mobile/core/storage/secure_storage_service.dart';
import 'package:vaxiil_mobile/features/business/data/organization_models.dart';

/// Shared consumer country filter: storage + headers + geo-country bootstrap.
class CountryScopeService {
  CountryScopeService({required SecureStorageService storage})
      : _storage = storage;

  final SecureStorageService _storage;
  DioClient? _dioClient;
  String _timezone = '';
  CountryBriefModel? _country;
  List<CountryBriefModel> _countries = [];
  Future<void>? _initPromise;

  CountryBriefModel? get country => _country;
  String? get countryId => _country?.id;
  String get isoCode2 => (_country?.isoCode2 ?? '').trim().toUpperCase();
  String get timezone => _timezone;
  List<CountryBriefModel> get countries => List.unmodifiable(_countries);

  void attachDio(DioClient client) {
    _dioClient = client;
  }

  Future<void> ensureTimezone() async {
    if (_timezone.isNotEmpty) return;
    try {
      _timezone = await FlutterTimezone.getLocalTimezone();
    } catch (_) {
      _timezone = '';
    }
  }

  Future<void> setCountry(CountryBriefModel country) async {
    _country = country;
    await _storage.writeMap(AppConstants.countryScopeStorageKey, {
      'id': country.id,
      'iso_code2': country.isoCode2.toUpperCase(),
      'name': country.name,
    });
  }

  Future<void> setCountryById(String countryId) async {
    for (final c in _countries) {
      if (c.id == countryId) {
        await setCountry(c);
        return;
      }
    }
  }

  void rememberCountries(List<CountryBriefModel> countries) {
    _countries = countries;
  }

  Future<void> ensureInitialized({
    List<CountryBriefModel> countries = const [],
    String? profileDefaultCountryId,
  }) {
    if (countries.isNotEmpty) {
      rememberCountries(countries);
    }
    return _initPromise ??= _bootstrap(
      profileDefaultCountryId: profileDefaultCountryId,
    );
  }

  void hydrateFromResolvedIso2(String? iso2) {
    if (_country != null || iso2 == null) return;
    final code = iso2.trim().toUpperCase();
    if (code.length != 2) return;
    for (final c in _countries) {
      if (c.isoCode2.toUpperCase() == code) {
        // ignore: discarded_futures
        setCountry(c);
        return;
      }
    }
  }

  Future<void> _bootstrap({String? profileDefaultCountryId}) async {
    await ensureTimezone();
    final stored = await _readStored();
    if (stored != null) {
      if (_countries.isNotEmpty) {
        for (final c in _countries) {
          if (c.id == stored.id ||
              c.isoCode2.toUpperCase() == stored.isoCode2.toUpperCase()) {
            await setCountry(c);
            return;
          }
        }
      }
      _country = stored;
      return;
    }

    if (profileDefaultCountryId != null && _countries.isNotEmpty) {
      for (final c in _countries) {
        if (c.id == profileDefaultCountryId) {
          await setCountry(c);
          return;
        }
      }
    }

    final detected = await _fetchGeoCountry();
    if (detected != null) {
      await setCountry(detected);
      return;
    }

    if (_countries.isNotEmpty) {
      await setCountry(_countries.first);
    }
  }

  Future<CountryBriefModel?> _fetchGeoCountry() async {
    final dio = _dioClient;
    if (dio == null) return null;
    try {
      await ensureTimezone();
      final response = await dio.get<dynamic>(
        AppConstants.organizationsGeoCountryPath,
        options: Options(
          headers: {
            if (_timezone.isNotEmpty) 'X-Timezone': _timezone,
          },
          validateStatus: (code) =>
              code != null && (code == 200 || code == 204),
        ),
      );
      final resolved = response.headers.value('x-resolved-country') ??
          response.headers.value('X-Resolved-Country');
      if (response.statusCode == 204 || response.data == null) {
        hydrateFromResolvedIso2(resolved);
        return null;
      }
      if (response.data is Map) {
        final country = CountryBriefModel.fromJson(
          Map<String, dynamic>.from(response.data as Map),
        );
        hydrateFromResolvedIso2(resolved);
        return country;
      }
    } catch (_) {}
    return null;
  }

  Future<CountryBriefModel?> _readStored() async {
    try {
      final map = await _storage.readMap(AppConstants.countryScopeStorageKey);
      if (map == null) return null;
      final id = map['id']?.toString() ?? '';
      final iso = map['iso_code2']?.toString() ?? '';
      if (id.isEmpty || iso.isEmpty) return null;
      return CountryBriefModel(
        id: id,
        isoCode2: iso.toUpperCase(),
        name: map['name']?.toString() ?? '',
      );
    } catch (_) {
      return null;
    }
  }
}
